src/main/lrsql/spec/admin.clj

```diff
@@ -17,8 +17,36 @@
 ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
 
 (s/def ::account-id ::c/primary-key)
-(s/def ::username string?)
-(s/def ::password string?)
+
+;; username and password are treated as simple non-empty strings
+(s/def ::username (s/and string? not-empty))
+(s/def ::password (s/and string? not-empty))
+
+;; Except on input validation!
+;; Minimum lengths
+(def u-min-len 7)
+(def p-min-len 10)
+
+(def digit-set (into #{} (map char (range 48 58))))  ; 0-9
+(def upper-set (into #{} (map char (range 65 91))))  ; A-Z
+(def lower-set (into #{} (map char (range 97 123)))) ; a-z
+(def special-set #{\! \@ \# \$ \% \^ \& \* \_ \- \+ \= \?})
+
+(s/def :lrsql.spec.admin.params/username
+  (s/and string?
+         #(>= (count %) u-min-len)
+         (partial re-matches #"^[a-zA-Z0-9]*$")))
+(s/def :lrsql.spec.admin.params/password
+  (s/and string?
+         #(>= (count %) p-min-len)
+         #(let [pass-set   (set %)
+                has-chars? (partial some pass-set)]
+            (and
+             (has-chars? digit-set)
+             (has-chars? upper-set)
+             (has-chars? lower-set)
+             (has-chars? special-set)))))
```

Comment:
I'm wondering if we should extract these additional checks into separate predicate functions. Spec's explain-data would then show the name of the refactored-out function instead of these anonymous functions, which may or may not be a good thing.

Reply:
I like that it is identical to what is in admin ui

Reply:
Fair enough.